src/functions.php

```diff
@@ -377,13 +386,132 @@ function upload(string $xml, $config)
+        foreach ($contact->telephony->number as $number) {
+            if ((isset($number->attributes()->quickdial) && $number->attributes()->quickdial >= 0)
+                || (isset($number->attributes()->vanity) && $number->attributes()->vanity != "")) {
+                $key = generateUniqueKey($number, $contact->carddav_uid);
+                $specialAttributes[$key] = $number->attributes();
+            }
+        }
+    }
+    return $specialAttributes;
+}
+
+
+/**
+ * Restore special attributes (quickdial, vanity) in given target phone book
+ *
+ * @param   SimpleXMLElement    $xmlTargetPhoneBook
+ * @param   array               $oldSpecialAttributes [] or map key => attributes
+ * @return  SimpleXMLElement    out phonebook with restored special attributes
+ */
+function mergePhoneNumberAttributes(SimpleXMLElement $xmlTargetPhoneBook, array $oldSpecialAttributes)
+{
+    if (!$oldSpecialAttributes) {
+        return $xmlTargetPhoneBook;
+    }
+    error_log("Restoring old special attributes (quickdial, vanity)".PHP_EOL);
+    foreach ($xmlTargetPhoneBook->phonebook->contact as $contact) {
+        foreach ($contact->telephony->number as $number) {
+            $key = generateUniqueKey($number, $contact->carddav_uid);
+            if (array_key_exists($key, $oldSpecialAttributes)) {
+                foreach (['quickdial','vanity'] as $attribute) {
+                    if (property_exists($oldSpecialAttributes[$key], $attribute)) {
+                        if (!property_exists($number->attributes(), $attribute)) {
+                            $number->addAttribute($attribute, (string)$oldSpecialAttributes[$key]->$attribute);
+                        } else {
+                            $number[$attribute] = (string)$oldSpecialAttributes[$key]->$attribute;
+                        }
+                    }
+                }
+            }
+        }
+    }
+    return $xmlTargetPhoneBook;
+}
+
+
+/**
+ * Build unique key with normalized phone number to lookup old phonebook attributes
+ * normalizing number means: remove all non-"+" and non-number characters like SPACE, MINUS, SLASH...
+ *
+ * @param   string  $number
+ * @param   string  $carddav_uid
+ * @return  string  format: {normalized-phone-number}@{vCard UUID}
+ */
+function generateUniqueKey(string $number, string $carddav_uid)
+{
+    return preg_replace("/[^\+0-9]/", "", $number)."@".$carddav_uid;
+}
+
+
+/**
+ * Check if special attributes already set (e.g., via CardDav extension 'X-FB-QUICKDIAL' / 'X-FB-VANITY')
+ *
+ * @param   SimpleXMLElement    $xmlPhonebook
+ * @return  boolean             true if any element already has a special attribute set
+ */
+function phoneNumberAttributesSet(SimpleXMLElement $xmlPhonebook)
+{
+    if (!property_exists($xmlPhonebook, "phonebook")) {
+        return false;
+    }
+    foreach ($xmlPhonebook->phonebook->contact as $contact) {
+        foreach ($contact->telephony->number as $number) {
```

Comment:
Could this break for contacts without numbers?

Reply:
No, then this loop is simply skipped and we return "false", as this contact does not have any quickdial / vanity attributes. We can keep this as is.